sql: connection to server consistently fails if started too soon

Found by trying to use the [testserver](https://github.com/cockroachdb/cockroach-go/tree/master/testserver)

The tests run a real cockroach binary to create a new single-node cluster and waits until `SHOW DATABASES` does not throw an error to ensure that we are properly initialized.

However, the testserver started failing recently with the query returning: `pq: database "information_schema" does not exist` at every attempt. Creating a new connection is successful.

A little bit of bisecting shows that the testserver was fine at `beta-20161201` but is failing with `beta-20161208`

see https://github.com/cockroachdb/cockroach-go/pull/22 for an easy way to reproduce:
```
~/cockroach/src/github.com/cockroachdb/cockroach-go/testserver $ go test ./... --cockroach-binary cockroach-beta-20161201.linux-amd64/cockroach 
ok  	github.com/cockroachdb/cockroach-go/testserver	0.537s

~/cockroach/src/github.com/cockroachdb/cockroach-go/testserver $ go test ./... --cockroach-binary cockroach-beta-20161208.linux-amd64/cockroach 
2016/12/13 09:03:38 Using custom cockroach binary: /store/cockroach-binaries/cockroach-beta-20161208.linux-amd64/cockroach
2016/12/13 09:03:38 process 1978 started: /store/cockroach-binaries/cockroach-beta-20161208.linux-amd64/cockroach start --logtostderr --insecure --port=0 --http-port=0 --socket=/tmp/cockroach-testserver608683055/.s.PGSQL.26257 --store=/tmp/cockroach-testserver608683055
2016/12/13 09:03:38 WaitForInit: pq: database "information_schema" does not exist
2016/12/13 09:03:38 WaitForInit: pq: database "information_schema" does not exist
<SNIP>
2016/12/13 09:03:43 WaitForInit: pq: database "information_schema" does not exist
2016/12/13 09:03:43 WaitForInit: pq: database "information_schema" does not exist
--- FAIL: TestRunServer (5.53s)
	testserver.go:149: pq: database "information_schema" does not exist
FAIL
FAIL	github.com/cockroachdb/cockroach-go/testserver	5.529s
```

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -519,34 +519,6 @@ func (s *Server) Start(ctx context.Context) error {
 		}))
 	})
 
-	if len(s.cfg.SocketFile) != 0 {
-		// Unix socket enabled: postgres protocol only.
-		unixLn, err := net.Listen("unix", s.cfg.SocketFile)
-		if err != nil {
-			return err
-		}
-
-		s.stopper.RunWorker(func() {
-			<-s.stopper.ShouldQuiesce()
-			if err := unixLn.Close(); err != nil {
-				log.Fatal(workersCtx, err)
-			}
-		})
-
-		s.stopper.RunWorker(func() {
-			pgCtx := s.pgServer.AmbientCtx.AnnotateCtx(context.Background())
-			netutil.FatalIfUnexpected(httpServer.ServeWith(s.stopper, unixLn, func(conn net.Conn) {
-				connCtx := log.WithLogTagStr(pgCtx, "client", conn.RemoteAddr().String())
-				if err := s.pgServer.ServeConn(connCtx, conn); err != nil && !netutil.IsClosedConnection(err) {
-					// Report the error on this connection's context, so that we
-					// know which remote client caused the error when looking at
-					// the logs.
-					log.Error(connCtx, err)
-				}
-			}))
-		})
-	}
-
 	// Enable the debug endpoints first to provide an earlier window
 	// into what's going on with the node in advance of exporting node
 	// functionality.
@@ -642,13 +614,39 @@ func (s *Server) Start(ctx context.Context) error {
 	log.Infof(ctx, "starting %s server at %s", s.cfg.HTTPRequestScheme(), unresolvedHTTPAddr)
 	log.Infof(ctx, "starting grpc/postgres server at %s", unresolvedListenAddr)
 	log.Infof(ctx, "advertising CockroachDB node at %s", unresolvedAdvertAddr)
-	if len(s.cfg.SocketFile) != 0 {
-		log.Infof(ctx, "starting postgres server at unix:%s", s.cfg.SocketFile)
-	}
-
 	s.stopper.RunWorker(func() {
 		netutil.FatalIfUnexpected(m.Serve())
 	})
+
+	if len(s.cfg.SocketFile) != 0 {
+		log.Infof(ctx, "starting postgres server at unix:%s", s.cfg.SocketFile)
+
+		// Unix socket enabled: postgres protocol only.
+		unixLn, err := net.Listen("unix", s.cfg.SocketFile)
+		if err != nil {
+			return err
+		}
+
+		s.stopper.RunWorker(func() {
+			<-s.stopper.ShouldQuiesce()
+			if err := unixLn.Close(); err != nil {
+				log.Fatal(workersCtx, err)
+			}
+		})
+
+		s.stopper.RunWorker(func() {
+			pgCtx := s.pgServer.AmbientCtx.AnnotateCtx(context.Background())
+			netutil.FatalIfUnexpected(httpServer.ServeWith(s.stopper, unixLn, func(conn net.Conn) {
+				connCtx := log.WithLogTagStr(pgCtx, "client", conn.RemoteAddr().String())
+				if err := s.pgServer.ServeConn(connCtx, conn); err != nil && !netutil.IsClosedConnection(err) {
+					// Report the error on this connection's context, so that we
+					// know which remote client caused the error when looking at
+					// the logs.
+					log.Error(connCtx, err)
+				}
+			}))
+		})
+	}
 
 	log.Event(ctx, "accepting connections")
 

